From: Ahmad Fatoum <a.fatoum@pengutronix.de>
Date: Thu, 3 Mar 2022 14:42:51 +0100
Subject: [PATCH] ARM: lxatac: register TLV decoder and print out serial

TLV for factory data is described in device tree. Register associated
decoder to match against it and to showcase use of decoded data, print
out device_serial_number.

As we do not yet employ deep probe for STM32MP1 platform, move board
driver registration to later point of time to be sure TLV device is
available. We separate out switch to have it happen before MAC setup.
This is not strictly necessary as networking isn't usable in barebox and
later on with DSA support, reset will happen in switch driver, but
splitting it up nevertheless improves readability.

Signed-off-by: Ahmad Fatoum <a.fatoum@pengutronix.de>
---
 arch/arm/boards/lxa-tac/board.c | 18 +++++++++++++++++-
 arch/arm/mach-stm32mp/Kconfig   |  2 ++
 2 files changed, 19 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boards/lxa-tac/board.c b/arch/arm/boards/lxa-tac/board.c
index 5403dda7680f..6e7d4bf77635 100644
--- a/arch/arm/boards/lxa-tac/board.c
+++ b/arch/arm/boards/lxa-tac/board.c
@@ -6,13 +6,29 @@
 #include <mach/bbu.h>
 #include <bootsource.h>
 #include <of.h>
+#include <tlv/tlv.h>
+#include <boards/lxa/tlv.h>
 
 static int tac_probe(struct device *dev)
 {
+	struct tlv_device *tlvdev;
+	int ret;
+
 	barebox_set_hostname("lxatac");
 
 	stm32mp_bbu_mmc_fip_register("mmc", "/dev/mmc1", BBU_HANDLER_FLAG_DEFAULT);
 
+	ret = lxa_tlv_v1_register();
+	if (ret)
+		return ret;
+
+	tlvdev = tlv_ensure_probed_by_alias("factory-data");
+	if (!IS_ERR_OR_NULL(tlvdev)) {
+		printf("tlvdev = %p\n", tlvdev);
+		dev_info(dev, "HW release: %s\n",
+			 dev_get_param(&tlvdev->dev, "device-hardware-release"));
+	}
+
 	return 0;
 }
 
@@ -26,4 +42,4 @@ static struct driver tac_board_driver = {
 	.probe = tac_probe,
 	.of_compatible = tac_of_match,
 };
-device_platform_driver(tac_board_driver);
+late_platform_driver(tac_board_driver);
diff --git a/arch/arm/mach-stm32mp/Kconfig b/arch/arm/mach-stm32mp/Kconfig
index e244bd392503..0155d3613197 100644
--- a/arch/arm/mach-stm32mp/Kconfig
+++ b/arch/arm/mach-stm32mp/Kconfig
@@ -28,10 +28,12 @@ config MACH_STM32MP15XX_DKX
 config MACH_LXA_MC1
 	select ARCH_STM32MP157
 	bool "Linux Automation MC-1 board"
+	select BOARD_LXA
 
 config MACH_LXA_TAC
 	select ARCH_STM32MP157
 	bool "Linux Automation TAC board"
+	select BOARD_LXA
 
 config MACH_SEEED_ODYSSEY
 	select ARCH_STM32MP157
